### ISSUE TITLE
Exclude Os that timeout from session refresh for a period of time

### ISSUE THREAD

**Issue by yondonfu** [livepeer/go-livepeer#1375]
**Is your feature request related to a problem? Please describe.**

B may encounter a HTTP timeout while waiting for a response from O. However, O may still respond quickly to B's `GetOrchestrator()` call. As a result, O could be consistently re-added to B's working set during broadcast session refreshes despite always timing out.

Furthermore, if O is still transcoding B's segment and B submits another segment to O, then B will immediately encounter an `OrchestratorBusy` error. The following example illustrates this problem:

1. B submits segment 1 to O1
2. B encounters an HTTP timeout
3. B starts a broadcast session refresh
4. O is added back to B's working set because it responds quickly to `GetOrchestrator()`
5. B submits segment 1 to O1
6. O1 is still working on segment 1 from step 1 so it returns `OrchestratorBusy`

**Describe the solution you'd like**

If B times out with an O, exclude that O from being re-added to the working set for a period of time in order to allow other Os to be added to the working set.

**Comment by j0sh** [comment]
Excluding Os from refreshes is a sensible workaround for mainnet or environments where there are a larger number of Os to choose from. For offchain private networks, we usually have fewer Os available.

Timing out and receiving `OrchestratorBusy` may be indicative of a larger problem, especially if this happens repeatedly with a limited set of Os. Simply excluding the O from the working set may be counterproductive if there isn't another O readily available to take its place. If the issue leading to the timeout is transient, it'll take longer for the B to latch back onto the O. The B is likely to see failures anyway: either it'll loop, get `OrchestratorBusy` and then exclude the O from the working set, or it'll preemptively exclude the O. Both have the net effect of shrinking the working set. Basically, the B "refreshing often" is a sign of other issues at play, especially if the set of Os is limited.

One way to reconcile this would be to define the exclusion period in terms of refreshes. Ideally, some of the goals would be:
* If there are more Os available, it gives the B an opportunity to run through the entire active set
* If there are fewer Os available, the exclusion period is shorter (or zero)
* If the B still refreshes frequently enough to be hitting OrchestratorBusy, that probably means there is some other problem that an exclusion period is not going to solve (eg, insufficient Os overall, poor connectivity on the B, etc)

The exclusion period could be defined like this:

```
exclusionPeriodInRefreshes = floor(activeSetSize  / 8)
```

The 8 is the maximum size of the B's working set  - the Os it has lined up for session selection. (The number 8 is defined in terms of [`(HTTP timeout / segment duration) * 2`](https://github.com/livepeer/go-livepeer/blob/77327a14158c87efe827339dbb0e9ce045f7341d/server/broadcast.go#L182-L183) : the maximum number of inflight segments, doubled in order to give us [room to refresh the set](https://github.com/livepeer/go-livepeer/blob/77327a14158c87efe827339dbb0e9ce045f7341d/server/broadcast.go#L71-L74) without dropping segments.) This is as opposed to the active set, which is the set of *all* Os available to the B during discovery. Of the active set, only 8 will be chosen at most for the working set.

Note there will be an interesting effect here, depending on the implementation. If `activeSetSize == 9` , and assuming an initial working set of 8, the B would typically refresh when 4 orchestrators are dropped from the working set. If the 4 dropped orchestrators need to be excluded (for one refresh), the B would have at most 5 Os after refreshing - which may in turn lead to another refresh after one more O is dropped. That refresh would likely result in the 4 dropped Os being re-added to the working set.

Following that line of thought, one possible approach is to attach a "refresh counter" as a prioritization mechanism. Something like:
* On exclusion, set `refreshPoint` to be `refreshCounter + exclusionPeriodInRefreshes`
* On refresh, non-excluded Os get `refreshCounter` as their  `refreshPoint` (or 0)
* Gather OrchestratorInfo responses (either up to a timeout or upon receiving responses from all Os), adding Os to the working set in order of `refreshPoint`.  Hence, Os with a lower `refreshPoint` get added before later (excluded) ones, but excluded Os would still be included if there is room in the working set. There should be a secondary ordering in terms of response time, the same as there is right now.
* After refresh, increment `refreshCounter`

This method would allow us to re-incorporate excluded Os immediately after exhausting the active set, without waiting for the next refresh.

This should be done per-stream, probably maintained within the broadcaster's session manager, with the `discovery` API updated to incorporate an exclusion list to `GetOrchestrators`.

Not sure how well this will perform in practice, but the behavior *feels* reasonable. It may be possible to get a better idea via simulation.

**Comment by j0sh** [comment]
After some testing and thinking about this, it'd be good to also exclude Os for other reasons in addition to timeouts:

* All HTTP-related errors
* Orchestrator-returned errors, such as OrchestratorBusy and OrchestratorCapped

In fact, our existing `shouldStopSession` errors seem to cover most of these cases:

https://github.com/livepeer/go-livepeer/blob/56a54051e13889fdd8bfb73011598f631fa4c9f1/server/broadcast.go#L479-L485

**Comment by kyriediculous** [comment]
# Exclude Orchestrators that timeout from the working set

## Abstract
Orchestrators can timeout during various points in the broadcasters lifecycle: 

- Initial discovery
- Discovery refresh loop (once every hour)
- When refreshing payment params if expired
- When transcoding
    - `OrchestratorBusy` error
    - HTTP Timeout (8 seconds)

For the first two cases the orchestrator will simply not be included in the `OrchestratorPool` that will make up the working set. 

The two latter cases are a little more tedious. We *do* remove the sessions when encountering a timeout here, but the session will be re-added when refreshing. 

Currently we refresh quite often as the maximum number of session the `BroadcastSessionManager` selects from the `OrchestratorPool` is limited to 8 orchestrators. The orchestrators to be included is decided by their response time to `GetOrchestratorInfo`, this response can be fast but the orchestrator can be very slow transcoding which leads to those orchestrators being reincluded when refreshing sessions*.

``` 
    // HTTP Timeout is 8 seconds
    // SegLen is 2 seconds
    // max segments in flight is 4
	maxInflight := common.HTTPTimeout.Seconds() / SegLen.Seconds()
	numOrchs := int(math.Min(poolSize, maxInflight*2))
```

_* refreshing sessions occurs when the session list falls below half of the original size eg. with 8 sessions in the session list we refresh when we have less than 4 sessions left._


## Motivation


Orchestrators that have fast ping times but are being dropped due to transcoding errors will get reincluded upon session refresh, this hurts latency and reliability. 

* Latency: having to refresh the session list, or reselecting an orchestrator adds to transcoding latency for a particular segment 

* Reliability: Subsequent retries in combination with HTTP timeouts can cause segments to be dropped. 


## Proposed Solution


**1. Add an exclusion list to the broadcaster.**

One approach could be to add a data structure to the `BroadcastSessionManager`, this would however be replicated for every new RTMP stream. This makes the space complexity grow by the amount of concurrent streams and the exclusion list would have to be re-established for each stream which partially reintroduces the problem we are trying to solve. 

Instead we could create a more persistent data structure on the `OrchestratorPool` itself to be used to filter out orchestrators when getting them from the `OrchestratorPool`. Active broadcasting sessions can send updates of orchestrators they want excluded over a channel.

This allows us to not have to _modify_ the existing API for `OrchestratorPool` but we can extend it instead and allows us to achieve the same (or perhaps better) effect while being slightly more efficient.  

**2. Prevent excluded orchestrators from being reinserted upon session refresh**

With an exclusion list being available on the `OrchestratorPool` and sessions being removed from the `BroadcastSessionManager` when timing out orchestrators can already be filtered out in the `OrchestratorPool.GetOrchestrators` call. Existing sessions upon refresh and newly created sessions will always have a list that excludes the necessary orchestrators. 

**3. Excluding orchestrators based on perceived transciency of the error (optional)**

Some transcoding errors might be more/less transient than others

- `OrchestratorBusy` indicates that the selected orchestrator is at capacity, but might actually be good at transcoding. 

- HTTP timeout indicates that the selected orchestrator can not return a satisfying result within 8 seconds. This can be due to bad capacity calibration on the orchestrator side or due to simply not having enough permanent capacity to transcode the requested renditions in real-time. 

This poses the question whether `OrchestratorBusy` is more transient than `HTTP Timeout` and if so if a higher penalty should be attached to less transient errors. 

## Implementation Tasks and Considerations


**OrchestratorPool Exclusion Feed**

For the data feed mechanism described above we can create a new channel for each `BroadcastSessionManager` in its constructor `NewSessionManager`. The channel can be closed when a stream finishes in `BroadcastSessionManager.cleanup()`.

We can extend the `OrchestratorPool` API with a `RegisterExclusionChannel` method to register new channels. When this method is called we will listen for updates on the channel until it is closed. 

The updates can be persisted in various forms: 

- In memory, exclude orchestrators `OrchestratorPool.GetOrchestrators`
- In DB (no need to worry about concurrency)
    - In a new table and fetch the results
    - In the existing orchestrators table by creating a new column eg. `exclusionTime` or `refreshPenalty` and adding a filter to `DBOrchFilter`

**How long to exclude orchestrators and how to quantify this**

There's a couple of options here. There is the proposal here to count this by number of refreshes. The proposal describes however that this would need to be tracked on a per-stream basis. The current solution described above where we track the exclusion list globally for all streams does slightly collide with this feature.

Instead we could also do this time-based, we know the segment length for a stream is `2 seconds` so regardless of the number of concurrent streams we can exclude the orchestrators by multiplications of 2 seconds (potentially based on the transiency of the error). 

Another option is to simply exclude orchestrators until the current active set size falls below a desired number and clear the exclusion list. 

## Testing Tasks and Considerations

- Assert that excluded transcoders are never selected

- Assert that excluded transcoders are never included in a session refresh

- Assert that excluded transcoders are never included in a new broadcast session

- Observe that the exclusion list is correctly updates based on the data feeds from the broadcast sessions

## Known Unknowns

- Should we exclude orchestrators based on transciency of the error ?
- Should we remove orchestrators from the exclusion list if they become deactive ? 
- What are the effects of increasing the maximum session list size to avoid refreshing as often? 

## Alternatives

- Tracking an exclusion list and refresh count penalty per stream: https://github.com/livepeer/go-livepeer/issues/1375#issuecomment-585389033

**Comment by j0sh** [comment]
> One approach could be to add a data structure to the BroadcastSessionManager, this would however be replicated for every new RTMP stream. This makes the space complexity grow by the amount of concurrent streams and the exclusion list would have to be re-established for each stream which partially reintroduces the problem we are trying to solve.

I'd appreciate it if we gave more consideration to the approach of managing the exclusion list via BroadcastSessionManager. First, let's acknowledge some drawbacks of that method:

* There will be increased per stream bookkeeping. However, the overall effect on memory usage feels comparatively small (proportional to the size of the orchestrator active set). Can this impact be quantified further?

* There will be a warm-up period for each stream. I imagine it's hard to quantify at the moment precisely how much of our initial failures on-chain result from such O errors. But we also have ways to mitigate the impact of these: retries, the single-segment-in-flight rule, and so forth.

Now, some reasons to consider a stream-level exclusion list:

* Introducing a node-level (global) exclusion list feels a bit premature. I have concerns about unintentionally making things more "twitchy" than they already are, especially in private/trusted network environments. For example, I'd prefer to look at ways to *dampen* the effect of possibly transient errors for otherwise reliable orchestrators. Doing this type of exclusion per-stream is one way of dampening such errors. With a global exclusion list, the working set becomes gradually smaller as the node gets busier, which has the potential for thundering herd problems that would decrease the throughput and reliability of the entire network.

* The semantic for an exclusion list does not necessarily have the total implication that "these orchestrators should not be worked with". The meaning could also be interpreted as, "we are already working with these orchestrators; please give me a fresh set that does not include these." Note that refreshes can and do incorporate orchestrators that the B is already working with for a stream.  A given refresh requesting N orchestrators could only be returning fewer "fresh" Os. This isn't really a problem in itself because you'll still be working with N orchestrators, but an exclusion mechanism could be useful for ramping up the size of the working set under some scenarios, eg to maximize O diversity or to expand the working set beyond some initial set. This could be useful for a highly parallel VOD workflow, for example.

* The two-second segment assumption is showing signs of strain under certain scenarios such as UGC where we don't always control user settings, or content type (low movement videos like a IP camera, screencasting a drawing or text editing), or VOD-type use cases. I'd guess that we're going to be making this flexible sooner rather than later. Introducing a global block list that makes a fixed assumption on the segment length makes this considerably more brittle.

All in all, a global exclusion list feels a bit like we're building ourselves into a corner at the moment. Would be great to consider a per-stream list.

**Comment by kyriediculous** [comment]
> There will be a warm-up period for each stream. I imagine it's hard to quantify at the moment precisely how much of our initial failures on-chain result from such O errors.

From what I have observed quite a bit considering not many orchestrators run enough computational capacity to transcode 1080p>1320p

Regardless of frequency I think it's important to note the high likelihood that if there's a high amount of early failures due to warm up time then the warm up time will be nearly equal for every stream being started due to the fact that we rely on the response time to `GetOrchestrator` to fill the working set. 

> For example, I'd prefer to look at ways to dampen the effect of possibly transient errors for otherwise reliable orchestrators. Doing this type of exclusion per-stream is one way of dampening such errors. With a global exclusion list, the working set becomes gradually smaller as the node gets busier, which has the potential for thundering herd problems that would decrease the throughput and reliability of the entire network.

The proposal above has quite a few differences with the eventual implementation in #1424. 
It combines a central suspension list with a refresh prioritisation mechanism. The working set will include suspended orchestrators if not enough non-suspended orchestrators are available. At this point they are removed from the suspension list and will not be re-added if they transcode properly again. 

So in an on-chain environment with default discovery we would never end up with a working set smaller than 8 if we refresh. 

We could decrease the sort of penalty through suspension by removing Os from the list after an arbitrary period of time, regardless of whether they have been selected due to not enough non-suspended O's being available or not. Would that decrease your level of concern? 

> I'd guess that we're going to be making this flexible sooner rather than later. Introducing a global block list that makes a fixed assumption on the segment length makes this considerably more brittle.

In #1424 we keep track of orchestrators and the time they were suspended at and no longer use segment length for anything. We use this timestamp in priority queue implemented as a min-heap when not enough non-suspended O's are available.

**Comment by yondonfu** [comment]
> We could decrease the sort of penalty through suspension by removing Os from the list after an arbitrary period of time, regardless of whether they have been selected due to not enough non-suspended O's being available or not.

We could actually still suspend an O for a certain # of refreshes as was originally suggested in [this comment](https://github.com/livepeer/go-livepeer/issues/1375#issuecomment-585389033). The one difference from the suggestion in the linked comment is that refreshes would be tracked on a global basis (i.e. each `GetOrchestrators()` could be treated as a refresh). This would mean that the higher the number of refreshes across all streams, the faster an O will be removed from the suspension list. The # of refreshes an O is suspended for could still be used to prioritize Os to pull off the suspension list if there are an insufficient number of non-suspended Os to fill the entire working set as described in the linked comment.

I see the primary benefit of the global suspension list (worth noting that this suspension list is in-memory and not persisted) as the reduced warm up period for each additional stream after the first. It seems to me that suspending an O for a certain # of refreshes and the prioritization mechanism for suspended Os when there are insufficient non-suspended Os for the working set largely addresses the concerns in [this comment](https://github.com/livepeer/go-livepeer/issues/1375#issuecomment-600782154) about using a global suspension list. Curious if @j0sh agrees/disagrees with this assessment and/or has additional concerns here though.

**Comment by j0sh** [comment]
> the eventual implementation in #1424. It combines a central suspension list with a refresh prioritisation mechanism. 

That approach to exclusion is nice. There might be an effect of making the ordering of selected orchestrators somewhat predictable, if the Os are consistent in their handling of OrchestratorInfo. This is probably not so different from the suggested [refresh based suspension](https://github.com/livepeer/go-livepeer/issues/1375#issuecomment-585389033) (and in fact it's simpler which is better), if the refresh-based exclusion period is scaled in order to cycle through the entire active set.

I'm still unsure of doing the suspension list globally. There are a couple side effects to be aware of here:

* Not all jobs are created equal, and an O that's unsuitable for the present job may be a good candidate for the next. The content itself may play a role as much as the transcoding preferences.

* The point at which an O is removed from exclusion becomes dependent on how busy the broadcaster is, and the size of the active set of Os. The current implementation means the set of Os could be cycled through very quickly, or very slowly, depending on the environment. This is one a form of hysteresis, but veers towards dampening *too much* or *not enough*. That is an effect I'd prefer to avoid.

Doing exclusions per-stream offers more predictability; it allows each stream to cycle through the active set as needed. New streams can latch onto Os quicker in the presence of transient errors. This is especially important with otherwise reliable / trusted networks, where per-stream exclusion would clearly help, but global exclusion would likely be detrimental to overall utilization.

> I see the primary benefit of the global suspension list (worth noting that this suspension list is in-memory and not persisted) as the reduced warm up period for each additional stream after the first.

Understand that a side effect of the global implementation is to reduce warm-up time. But note that "reducing warm up time" wasn't necessarily one of the original goals here. A concerted effort to reduce warm-up time might also take previous successes into account, rather than just failures. That is also something that feels more appropriate for onchain mode, whereas per-stream exclusion is still useful offchain.

**Comment by yondonfu** [comment]
> Not all jobs are created equal, and an O that's unsuitable for the present job may be a good candidate for the next

Yeah great point. But, at the moment, it seems that of the errors that an O could be suspended for, the only one that would be impacted by job type is a timeout error since O could timeout for a more complex job, but not timeout for a less complex job. `OrchestratorCapped` is based on the # of streams that an O is transcoding [1]. `OrchestratorBusy` only occurs if B inadvertently bypasses the one segment in flight rule which I don't think should happen if B is using a suspension list (regardless of whether it is global or per-stream) since B would not send a segment to an O that timed out recently.

While I do see why a timeout for one job may not necessarily make an O inadequate for another job due to a different in job type, it also feels reasonable to allow B to cycle through the rest of the active set first to give it the opportunity to find a better O (and if there are insufficient non-suspended Os B will still be able to latch onto this O anyway).

[1] This could change if we ever change O caps to factor in job complexity.

Regarding your point in #1433 (moving discussion of this point here since it feels more relevant to this thread):

> Note that this is yet another reason that a per-stream exclusion list makes more sense (#1375) : a stream requesting an older supported format may work with a given orchestrator, while another stream on the same B using a newer format may not work with the same orchestrator.

Format support and more generally transcoding preference supports seems like something that can be determined at discovery time instead of segment submission time. For example, B's GetOrchestrator request could include the transcoding preferences and O's could respond with an error during discovery if the preferences are not supported (i.e. a format is not supported). So, we could filter out Os based on transcoding preferences during discovery instead of using a suspension list.

> This is especially important with otherwise reliable / trusted networks, where per-stream exclusion would clearly help, but global exclusion would likely be detrimental to overall utilization.

Would you mind elaborating a bit more on why global suspension would likely be detrimental in a reliable/trusted network? I see the pros of global suspension in an open/untrusted network, but I'm not as clear on the cons of it in a reliable/trusted network. Perhaps if I had a better understanding I'd have a clearer view of why per-stream suspension might be better here.

By detrimental to overall utilization do you mean prematurely routing streams to different Os instead of making sure that the Os that B already is aware of are fully utilized (i.e. there may be intermittent errors due to being at capacity, but whenever they are not capacity we want B to have them transcode a segment)? If there are other Os available is this so bad?

My understanding is that there is a worry of the thundering herd problem where a single O is overwhelmed by streams because they are routed to it simultaneously. If we tend to stick with an O that has a transient error, perhaps because it is at capacity, wouldn't we be running into a similar problem? In this case, it seems reasonable to give the O a "longer break" in the form of a suspension period as opposed to allowing streams that are ingested by B within a short time duration to all be routed to that O.

I also understand that a downside of the global suspension list can dampen hysteresis too much or too little. But, is that really that big of a concern if there other available Os and if suspended Os can be used if there are insufficient non-suspended Os?

I discussed this global vs. per-stream suspension list topic with @kyriediculous today and the current plan is to:

- Update the `OrchestratorPool` API in #1424 so that the caller passes in the suspension list to `GetOrchestrators` (as per [this suggestion](https://github.com/livepeer/go-livepeer/issues/1375#issuecomment-585389033))
- The above will allow the caller to determine how the suspension list is instantiated
- The suspension list will remain global for the time being, but changing it to be per-stream should just be a matter of instantiating a separate suspension list per `BroadcastSessionsManager` instead of instantiating a single suspension list that is used by each `BroadcastSessionsManager`
- Do some testing on mainnet with the global suspension list
- Remain open to the possibility of a per-stream suspension list (the changes required should be small after the `OrchestratorPool` API update) as we better understand why a per-stream suspension list might be better when considering its use in reliable/trusted networks

**Comment by j0sh** [comment]
TL/DR: Two primary areas of concern with global O exclusion.

(This is probably the last I'll write on the topic; will leave the rest up to @yondonfu and @kyriediculous . The design proposed in https://github.com/livepeer/go-livepeer/issues/1375#issuecomment-606307566 seems to be on the right track , but I haven't reviewed #1435 . )

* Re-latching back onto an excluded O is dependent on the B's activity level. This is unpredictable, and I believe we should err on the side of "quicker latching".

* Issues with a particular stream or segment, rather than the O itself, can have knock-on effects on how unrelated streams use the network. This leads to (un-)balanced utilization which has steeper consequences on private networks.

> of the errors that an O could be suspended for, the only one that would be impacted by job type is a timeout error since O could timeout for a more complex job, but not timeout for a less complex job. OrchestratorCapped is based on the # of streams that an O is transcoding [1].

OrchestratorCapped (and OrchestratorBusy / timeouts) are good examples of why I like per-stream exclusion. OrchestratorCapped is by definition transient [1]: a specific stream might not want to re-acquire an O that's capped, but another stream a few moments later might have no such issue if capacity frees up. Otherwise latching back on to the O is dependent on the B's activity level, which may be unpredictable. (More on latching later.)

For private networks where reliability is assumed, timeouts or OrchestratorBusy could indicate issues with a particular stream / segment as much as a problem with the O itself. In fact, the way to distinguish between the two is by observing the rate of errors for both the stream (retries) and the O. By immediately throttling an O in its entirety, we not only lose that visibility on the O side (aside from other streams it may already be working with, which is not a given), it may have a detrimental effect on other streams by unnecessarily reducing O availability, especially across retries of a faulty segment / heavy job. I think this will be too twitchy.

[1] Aside: I *believe* that the capacity cap should be [checked](https://github.com/livepeer/go-livepeer/blob/2752da4a394b5623dce06542d6e64848b121efd8/server/rpc.go#L286) as part of the OrchestratorInfo request, so it might not be completely necessary to have `OrchestratorCapped` in the exclusion criteria. However, I distinctly recall seeing the need for it when doing utilization testing a few months ago. Possible that the OrchestratorInfo capacity check doesn't actually work as it's supposed to.

> it also feels reasonable to allow B to cycle through the rest of the active set first to give it the opportunity to find a better O
>
> it seems reasonable to give the O a "longer break" in the form of a suspension period
>
> By detrimental to overall utilization do you mean prematurely routing streams to different Os instead of making sure that the Os that B already is aware of are fully utilized

Yeah so that's one of the concerns: latching back on to an excluded O becomes dependent on the activity level of the B (and the size of the O active set to an extent). On public networks, O utilization is a lesser concern, but more so on private networks. Os that are available-but-not-used may have a cost, both in keeping the O available, and in increased load on the rest of the Os, which in turn increases latency, reduces the margin of error, etc. This is more acute with smaller setups which may otherwise still be completely reasonable (eg, 3-4 Os). In the worst case, an O being excluded could mean the additional marginal load on another O reaches the breaking point of exclusion.

The newest implementation limits exclusions based on the number of refreshes which is nice because it also makes the exclusions a function of the O set size. Under normal conditions, a stream should hardly ever need to call refresh. If many streams are calling refresh frequently, that is usually a sign something else is going on. Now, in those cases, it *feels* a little more robust for each stream to have the opportunity to converge on a satisfactory set of orchestrators independently. Doing this globally seems a little more prone to inducing herd-like behavior - which is *usually* undesirable. I don't have any hard data on this right now, but it feels like the better approach from my experience testing various scenarios.

Re-latching can happen as soon as a new stream comes online, and the stream itself can cycle through the active set at its own pace. If there's a problem with a particular stream, then the problem is isolated to that one stream; it reduces the blast radius of globally excluding the O for all other streams.

> Format support and more generally transcoding preference supports seems like something that can be determined at discovery time instead of segment submission time

Absolutely, discovery is the ideal place for capability checking. https://github.com/livepeer/go-livepeer/pull/1433 is an extremely coarse example of that [1]. However, robust capability-driven discovery is going to need some significant upfront design and thinking [2].

On the other hand, returning a transcode error on an unsupported format is something we need to do anyway. Enabling blended networks is a nice side effect of per-stream exclusion without further discovery improvements [3], just as reducing warm-up time is a side effect of global exclusion on public networks. The former can be useful for both public / private networks [4], while the latter only seems useful for the public network.

[1]  Whether we should actually expand on https://github.com/livepeer/go-livepeer/pull/1433 into something less coarse is open for discussion in that PR.

[2] Another way to do it: rather than include the desired configuration in B's GetOrchestrator request (which implies *sending* the information, when it may not always be available upfront, eg during caching), the O sends down everything it supports in the OrchestratorInfo response (eg, a 'menu' or 'catalog'). The B can then filter as necessary (or not); eg, it can cache the capabilities for each O, and for each stream, only send requests to orchestrators that claim to support the stream's format. That allows for gathering a view of all capabilites across the network, potentially with pricing attached to each capability. Anyway, just one example: it's a fairly deep topic with a lot of design questions and implications.

[3] Well, "no further discovery changes" aside from #1433 which ironically does *not* allow for blended networks before/after the PR but is still necessary for correctness.

[4] We can't expect all nodes to be upgraded in tandem even on a private network. Incremental roll-outs, etc.

**Comment by kyriediculous** [comment]
- Re-latching back onto an excluded O is dependent on the B's activity level. This is unpredictable, and I believe we should err on the side of "quicker latching".

If the activity level is low and the current set of orchestrators is sufficient there is no need to latch on to old orchestrators and gives others the chance to perform even better. A known orchestrator is not necessarily better than an unknown orchestrator especially if we are certain the known orchestrator can only handle a job with requirements below the one it errored out for. 

> Re-latching can happen as soon as a new stream comes online, and the stream itself can cycle through the active set at its own pace. If there's a problem with a particular stream, then the problem is isolated to that one stream; it reduces the blast radius of globally excluding the O for all other streams.

This assumes that an error encountered for one stream doesn't  impact other streams, which I don't agree with in the case of Capped and Timeout. They will impact other streams for their duration and while the goal of this project is not necessarily to improve warm up time, the warm-up time directly affects reliability. As long as there are no better alternatives on the table to reduce warm up time in the short term and since switching from global to per stream suspension is quite trivial at this point I'd be in favor of keeping the global suspension list. 

Then again there is the opportunity for a per stream suspension list as a less coarse mechanism for dealing with different formats as per my comment in [#1433 ](https://github.com/livepeer/go-livepeer/pull/1433#issuecomment-608632421)


 > latching back on to an excluded O becomes dependent on the activity level of the B (and the size of the O active set to an extent). On public networks, O utilization is a lesser concern, but more so on private networks. Os that are available-but-not-used may have a cost, both in keeping the O available, and in increased load on the rest of the Os, which in turn increases latency, reduces the margin of error, etc. This is more acute with smaller setups which may otherwise still be completely reasonable (eg, 3-4 Os). In the worst case, an O being excluded could mean the additional marginal load on another O reaches the breaking point of exclusion.

With small active set sizes and small working set sizes the penalty will be only 1 refresh. So in essence this mimics a per-stream list as every subsequent stream will remove it from suspension again for all streams, allowing _all_ streams to now latch back on to that O which allows even faster re-latching than per-stream suspension.

**Comment by yondonfu** [comment]
After a discussion with @j0sh, it looks like the coarse wire protocol versioning scheme in #1433 will be replaced with a modification to the `net.SegData` message sent to Os in order to force older Os that do not support a format (i.e. MP4) to fail early. A per-stream suspension list will exclude older Os that do not support the requested format for a stream while allowing older Os that do support the requested format to be used for other streams. This will be helpful as a temporary filtering mechanism at segment submission time prior to the introduction of capability discovery.

Given the per-stream suspension list's utility in the case mentioned above and the possibility of addressing the warm up time of additional streams in other more robust/complete ways (possibilities include redundancy, reputation/satisfaction score tracking or even a global suspension list for certain types of errors to complement the per-stream suspension lists) I think we should move forward with a per-stream suspension list.
